onnxruntime/core/framework/session_state.h

```diff
@@ -232,7 +231,8 @@ class SessionState {
       std::unordered_map<onnxruntime::NodeIndex, std::unordered_map<std::string, std::unique_ptr<SessionState>>>;
   SubgraphSessionStateMap subgraph_session_states_;
 
-  onnxruntime::concurrency::ThreadPool* thread_pool_ = nullptr;
+  //It could be NULL
+  concurrency::ThreadPool* const thread_pool_;
```

Comment:
Should we set it up so that it never returns null, unless there's a reason to allow that?

Reply:
For executing as single threaded.

Reply:
Seems like there are two choices here:
1) The thread pool implementation handles running single threaded if it's created with zero threads (does what it does currently and executes the provided function on the calling thread).
2) All code that gets a threadpool from session state has to check for nullptr and have one path that uses the threadpool and one path that doesn't

The first seems simpler to me, but either are possible ways to execute single threaded.

---
In reply to: [314569846](https://github.com/microsoft/onnxruntime/pull/1609#discussion_r314569846) [](ancestors = 314569846)